flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarUnorderedPartitionSplitReader.java

```diff
@@ -137,16 +137,18 @@ protected void startConsumer(PulsarPartitionSplit split, Consumer<byte[]> consum
 
             // Redeliver unacknowledged messages because of the message is out of order.
             consumer.redeliverUnacknowledgedMessages();
-        } else {
-            initialStartPosition(split, consumer);
         }
     }
 
     public PulsarPartitionSplitState snapshotState(long checkpointId) {
-        TxnID txnID = PulsarTransactionUtils.getId(uncommittedTransaction);
-        this.uncommittedTransaction = newTransaction();
         PulsarPartitionSplitState state = new PulsarPartitionSplitState(registeredSplit);
-        state.setUncommittedTransactionId(txnID);
+
+        // Avoiding NP problem when Pulsar don't get the message before Flink checkpoint.
+        if (uncommittedTransaction != null) {
```

Comment:
Please add `@Nullable` to `uncommittedTransaction` field declaration that helps IDE to warn us for such issues.